langwatch/src/optimization_studio/components/workflow/NewWorkflowModal.tsx

```diff
@@ -1,14 +1,78 @@
-import { useEffect, useState } from "react";
-import { Button, Separator, Grid, HStack, Text } from "@chakra-ui/react";
+import { useEffect, useState, useRef, useCallback } from "react";
+import {
+  Button,
+  Separator,
+  Grid,
+  HStack,
+  Text,
+  Box,
+  VStack,
+} from "@chakra-ui/react";
 import { Dialog } from "../../../components/ui/dialog";
 import { WorkflowCard } from "./WorkflowCard";
-import { ChevronLeft, File } from "react-feather";
+import { ChevronLeft, File, Upload } from "react-feather";
 import { NewWorkflowForm } from "./NewWorkflowForm";
 import type { Workflow } from "../../types/dsl";
 import { TEMPLATES } from "../../templates/registry";
+import { toaster } from "../../../components/ui/toaster";
+import { workflowJsonSchema } from "../../types/dsl";
+
+/** Maximum allowed file size for workflow imports (5MB) */
+const MAX_WORKFLOW_FILE_SIZE = 5 * 1024 * 1024;
 
 type Step = { step: "select" } | { step: "create"; template: Workflow };
 
+/**
+ * Hook for handling file drag and drop functionality
+ * @param onFileSelect Callback when a file is selected
+ * @returns Object containing drag state and event handlers
+ */
+function useFileDrop(onFileSelect: (file: File) => void) {
+  const [isDragging, setIsDragging] = useState(false);
+  const fileInputRef = useRef<HTMLInputElement>(null);
+
+  const handleDragOver = useCallback((e: React.DragEvent) => {
+    e.preventDefault();
+    setIsDragging(true);
+  }, []);
+
+  const handleDragLeave = useCallback(() => {
+    setIsDragging(false);
+  }, []);
+
+  const handleDrop = useCallback(
+    (e: React.DragEvent) => {
+      e.preventDefault();
+      setIsDragging(false);
+      const file = e.dataTransfer.files[0];
+      if (file) onFileSelect(file);
+    },
+    [onFileSelect]
+  );
+
+  const handleClick = useCallback(() => {
+    fileInputRef.current?.click();
+  }, []);
+
+  const handleFileInputChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const file = e.target.files?.[0];
+      if (file) onFileSelect(file);
```

Comment:
I think this is where the error would be thrown. What happens if this errors, is it caught?